eclair-core/src/main/scala/fr/acinq/eclair/payment/send/MultiPartPaymentLifecycle.scala

```diff
@@ -216,18 +198,21 @@ class MultiPartPaymentLifecycle(nodeParams: NodeParams, cfg: SendPaymentConfig,
       } else {
         stay using d.copy(pending = pending)
       }
+
+    case Event(Stop, d: PaymentSucceeded) => myStop(d.sender, Right(cfg.createPaymentSent(d.preimage, d.parts)))
+
+    case Event(_: RouteResponse, _) => stay
+    case Event(_: Status.Failure, _) => stay
   }
 
   onTransition {
     case _ -> PAYMENT_ABORTED => nextStateData match {
-      case d: PaymentAborted if d.pending.isEmpty =>
-        myStop(d.sender, Left(PaymentFailed(id, paymentHash, d.failures)))
+      case d: PaymentAborted if d.pending.isEmpty => self ! Stop
```

Comment:
I introduced this instead of directly calling `myStop`, because when calling `myStop` directly, the FSM doesn't terminate during the tests. It feels like calling `FSM.stop` inside a transition has a slightly different result than doing it from a known state. I haven't investigated that much, but if someone has seen this behavior before I'm interested in what they learnt.

Reply:
Maybe using transition for this isn't a good pattern. How about:

```scala
def gotoAbortedOrStop(d: PaymentAborted): State = {
    if (d.pending.isEmpty) {
      myStop(d.sender, Left(PaymentFailed(id, paymentHash, d.failures)))
    } else
      goto(PAYMENT_ABORTED)
  }
```

Reply:
Do you really dislike the pattern of sending ourselves a message in the transition?
I quite like this centralized transition, it makes sure we cannot shoot ourselves in the foot by doing a transition without using the `gotoAbortedOrStop` method.

Reply:
The more I think about it, the less I like the pattern of intercepting and changing the next state from within a transition. Sending a message to ourself is fine, it's just the overall pattern that I don't like.

Reply:
Really? I think it makes sense that we can only stop from terminal states (`PAYMENT_ABORTED` and `PAYMENT_SUCCEEDED`), what I find lacking is a hook when arriving into a new state, which is somewhat replaced by the `onTransition` + sending events to ourselves trick. Let me know if you find a better solution.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1439/commits/34204e1ca5e8e65b5683672708aadbddb04f7b60